src/groupeddataframe/grouping.jl

```diff
@@ -8,11 +8,17 @@ Not meant to be constructed directly, see `groupby`.
 """
 struct GroupedDataFrame{T<:AbstractDataFrame}
     parent::T
-    cols::Vector{Int}    # columns used for grouping
-    groups::Vector{Int}  # group indices for each row
-    idx::Vector{Int}     # indexing vector when grouped by the given columns
-    starts::Vector{Int}  # starts of groups
-    ends::Vector{Int}    # ends of groups
+    cols::Vector{Int}         # columns used for grouping
+    groups::Vector{Int}       # group indices for each row
+    idx::Vector{Int}          # indexing vector when grouped by the given columns
+    starts::Vector{Int}       # starts of groups
+    ends::Vector{Int}         # ends of groups
+    values::Vector{<: Tuple}  # values of grouped columns
+end
+
+function GroupedDataFrame(parent, cols, groups, idx, starts, ends)
+    values = [Tuple(parent[idx[s], cols]) for s in starts]
```

Comment:
I will return to an earlier discussion. As @nalimilan suggested - this can be very expensive if we have a lot of small groups. Maybe for the time being we can leave-out the `values` field and only calculate it on the fly (which should not be problematic unless we reuse `gd` many times for iteration later, which I guess is a rare case, and if someone really wants it you can always store `keys` to get this)?

The problem with lazy creation of `values` field is that we cannot do it without materializing as `Tuple`s inside it might have heterogeneous type.

Reply:
Also - if we decide to materialize `values` probably `Vector` is not optimal, as it would be better to use `Dict`. The reason is that `keys` is an operation that you do only once (and then `Vector` is useful), but `findfirst` operation is `O(n)`, so in practice the lookup in `getindex` will be slow with `Vector`.

Finally, no matter if we use `Tuple` or `Dict` unless we improve the code here we will normally get a bad type inference in corner cases with `Missing` (which are to be expected in DataFrames.jl). What I mean is:
```
julia> [(1,missing), (missing, 1), (1,2)]
3-element Array{Tuple{Any,Any},1}:
 (1, missing)
 (missing, 1)
 (1, 2)
```
and
```
julia> Dict((1,missing)=>1, (missing, 1)=>2, (1,2)=>3)
Dict{Tuple{Any,Any},Int64} with 3 entries:
  (1, 2)       => 3
  (missing, 1) => 2
  (1, missing) => 1
```

In summary, we can do two things:
1. leave out materializing `values` and warn users that the `getindex` and `keys` operation is currently lazy and might be slow in hot loops (so they should themselves materialize the groups and use integer indexing which is fast for such use-cases)
2. work out the best design of `values` related field (that will not have performance impact if this filed is not needed, has correct eltype identification and is fast in the relevant use-cases)

I would opt for 1 in this PR (as this is functionality related) and leave option 2 for a separate PR (for @jlumpe to submit - if you wish; alternatively I can think about it separately). The reason is that then we will already have a working functionality in DataFrames.jl and only performance optimization (which is non-breaking, but potentially it will be a longer discussion about what is the best design here), will be left for later.